src/mlpack/methods/ann/loss_functions/binary_cross_entropy_loss_impl.hpp

```diff
@@ -34,11 +34,13 @@ BCELoss<InputDataType, OutputDataType>::Forward(
 {
   typedef typename PredictionType::elem_type ElemType;
 
-  ElemType loss = -arma::accu(target % arma::log(prediction + eps) +
+  ElemType lossSum = -arma::accu(target % arma::log(prediction + eps) +
       (1. - target) % arma::log(1. - prediction + eps));
+
   if (reduction)
-    loss /= prediction.n_elem;
-  return loss;
+    return lossSum;
+
+  return lossSum / target.n_elem;
```

Comment:
Shouldn't that be batch-size, same as in [cosine_embedding_loss_impl.hpp](https://github.com/mlpack/mlpack/pull/3144/files#diff-d525bfb306bbe31607580cff9c6d8721ff0a29959dcb4c895b449050b23bac32).

Reply:
In mean reduction, I'm dividing loss by the number of elements in output (Pytorch also does the same). The implementation of Cosine Embedding Loss in mlpack & Pytorch are slightly different. Pytorch takes `y` as target & in mlpack `y` is a hyperparamater. So, to match the results with Pytorch, the loss is divided by batchSize ( which is equivalent to dividing by no. of elements in y). However,  I'm not sure if the calculation of batchSize is correct. It's taken as `prediction.elem  / prediction.n_cols` which is equivalent to `prediction.n_rows * no. of slices` but don't we assume that each column represents a datapoint & not each row?

Reply:
Yes, each column represents another sample.